Sources/SwiftDocCUtilities/Action/Actions/Convert/ConvertAction.swift

```diff
@@ -329,15 +282,10 @@ public struct ConvertAction: AsyncAction {
             workingDirectory: temporaryFolder,
             fileManager: fileManager)
 
-        // An optional indexer, if indexing while converting is enabled.
-        var indexer: Indexer? = nil
-        
-        let bundleIdentifier = converter.firstAvailableBundle()?.identifier
-        if let bundleIdentifier {
-            // Create an index builder and prepare it to receive nodes.
-            indexer = try Indexer(outputURL: temporaryFolder, bundleIdentifier: bundleIdentifier)
-        }
+        let indexer = try Indexer(outputURL: temporaryFolder, bundleIdentifier: bundle.identifier)
 
+        let context = try DocumentationContext(bundle: bundle, dataProvider: dataProvider, diagnosticEngine: diagnosticEngine, configuration: configuration)
```

Comment:
> Since we're refactoring this top level code now, I'd suggest splitting this up into two calls:
> 
> * `let context = DocumentationContext()`
> * `context.register`

No, at least not on the context itself. That would essentially put us back where we started with a two-step creation:

```swift
let workspace = DocumentationWorkspace()
let context = DocumentationContext(dataProvider: workspace, ...)
// other code
workspace.register(...)
```

There are three main problems with this pattern:
- The context needs to be fully mutable since its won't have its "initial" values until `register` is called.
- It's possible to keep modifying the context up until `register` but also _after_. Sometimes this led to bugs where a property was modified after registration and thus had no effect
- There's nothing preventing `register` from running more than once. 

----- 

However, one idea that I've been thinking about for the is to move all the `register` work into a new `DocumentationContextBuilder` type with
```swift
consuming func build() -> DocumentationContext {}
```

This would allow us to move the registration code out of `DocumentationContext` and doesn't require making/keeping the context fully mutable, while still preventing bugs where the builder is modified after creating the context and preventing `build()` from being called more than once (because the builder is consumed and can no longer be used after returning the built context).

One of the reasons why I've keep the new context initializer `package` access is so that we can make that type of changes without needing even more rounds of deprecations.

Reply:
That said, one idea that I've been thinking about for the is to move all the `register` work into a new `DocumentationContextBuilder` type with
```swift
consuming func build() -> DocumentationContext {}
```

This would allow us to move the registration code out of `DocumentationContext` and doesn't require making/keeping the context fully mutable, while still preventing bugs where the builder is modified after creating the context and preventing `build()` from being called more than once (because the builder is consumed and can no longer be used after returning the built context).

One of the reasons why I've keep the new context initializer `package` access is so that we can make that type of changes without needing even more rounds of deprecations. 

Unfortunately I think that it would be challenging to make that change before 6.2 and maintain compatible with the two-step workspace-style setup without duplicating the existing registration code.

Reply:
Ok makes sense - I didn't consider mutability or the possibility of someone calling `register` twice.